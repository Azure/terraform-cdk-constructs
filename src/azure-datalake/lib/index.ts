export * from "./filesystem";
export * from "./path";
export * from "./datalake";
