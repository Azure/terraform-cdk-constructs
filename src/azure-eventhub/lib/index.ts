export * from "./authorization";
export * from "./eventhubcluster";
export * from "./consumer";
export * from "./instance";
export * from "./kusto-connection";
export * from "./namespace";
