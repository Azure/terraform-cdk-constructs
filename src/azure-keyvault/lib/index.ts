export * from './certificate'
export * from './key'
export * from './policy'
export * from './secret'
export * from './vault'

