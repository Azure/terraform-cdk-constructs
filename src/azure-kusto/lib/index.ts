export * from "./cluster";
export * from "./compute-specification";
export * from "./database";
