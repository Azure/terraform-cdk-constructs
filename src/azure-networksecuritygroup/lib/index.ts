export * from './preconfigured-rules';
export * from './securitygroup';
