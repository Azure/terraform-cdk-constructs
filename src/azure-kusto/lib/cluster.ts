import { KustoCluster } from "@cdktf/provider-azurerm/lib/kusto-cluster";
import * as cdktf from "cdktf";
import { Construct } from "constructs";
import {
  ComputeSpecification,
  IComputeSpecification,
} from "./compute-specification";
import { Database, DatabaseProps } from "./database";
import { Group } from "../../azure-resourcegroup";
import { AzureResource } from "../../core-azure/lib/index";

export interface ClusterProps {
  /**
   * The Azure Resource Group in which to create the Kusto Cluster.
   */
  readonly rg: Group;
  /**
   * The name of the Kusto Cluster to create.
   * Only 4-22 lowercase alphanumeric characters allowed, starting with a letter.
   */
  readonly name: string;
  /**
   * The SKU of the Kusto Cluster. All the allowed values are defined in the ComputeSpecification class.
   * @default devtestExtraSmallDv2
   */
  readonly sku?: IComputeSpecification;
  /**
   * The node count for the cluster.
   * @default 2
   */
  readonly capacity?: number;
  /**
   * The type of Managed Service Identity.
   * @default "SystemAssigned"
   */
  readonly identityType?: string;
  /**
   * A list of User Assigned Managed Identity IDs to be assigned to this Kusto Cluster.
   */
  readonly identityIds?: string[];
  /**
   * Is the public network access enabled?
   * @default true
   */
  readonly publicNetworkAccessEnabled?: boolean;
  /**
   * Specifies if the cluster could be automatically stopped.
   * (due to lack of data or no activity for many days).
   * @default true
   */
  readonly autoStopEnabled?: boolean;
  /**
   * Specifies if the streaming ingest is enabled.
   * @default true
   */
  readonly streamingIngestionEnabled?: boolean;
  /**
   * Specifies if the purge operations are enabled.
   * @default false
   */
  readonly purgeEnabled?: boolean;
  /**
   * Specifies if the purge operations are enabled. Based on the SKU, the number of zones allowed are different.
   * @default true
   */
  readonly enableZones?: boolean;
  /**
   * The minimum number of allowed instances. Must between 0 and 1000.
   */
  readonly minimumInstances?: number;
  /**
   * The maximum number of allowed instances. Must between 0 and 1000.
   */
  readonly maximumInstances?: number;
  /**
   * A mapping of tags to assign to the Kusto.
   */
  readonly tags?: { [key: string]: string };
}

export class Cluster extends AzureResource {
  readonly kustoProps: ClusterProps;
  public readonly id: string;
  public readonly resourceGroupName: string;
  public readonly uri: string;

  constructor(scope: Construct, id: string, kustoProps: ClusterProps) {
    super(scope, id);
    this.kustoProps = kustoProps;
    this.resourceGroupName = kustoProps.rg.Name;

    /**
     * Define default values.
     */
    const sku = kustoProps.sku || ComputeSpecification.devtestExtraSmallDv2;
    const enableZones = kustoProps.enableZones || true;

    const defaults = {
      publicNetworkAccessEnabled: kustoProps.publicNetworkAccessEnabled || true,
      autoStopEnabled: kustoProps.autoStopEnabled || true,
      streamingIngestionEnabled: kustoProps.streamingIngestionEnabled || true,
      purgeEnabled: kustoProps.purgeEnabled || false,
      zones: enableZones ? sku.availibleZones : [],
      sku: {
        name: sku.skuName,
        capacity: kustoProps.capacity || 2,
      },
      identity: {
        type: "SystemAssigned",
        identityIds: [],
      },
    };

    /**
     * Create Kusto Cluster resource.
     */
    const azurermKustoCluster = new KustoCluster(this, "Kusto", {
      ...defaults,
      name: kustoProps.name,
      location: kustoProps.rg.Location,
      resourceGroupName: kustoProps.rg.Name,
      tags: kustoProps.tags,
    });

    if (kustoProps.identityType) {
      azurermKustoCluster.addOverride("identity", {
        type: kustoProps.identityType,
        identityIds: kustoProps.identityIds,
      });
    }

    if (kustoProps.minimumInstances && kustoProps.maximumInstances) {
      azurermKustoCluster.addOverride(
        "minimum_instances",
        kustoProps.minimumInstances,
      );
      azurermKustoCluster.addOverride(
        "maximum_instances",
        kustoProps.maximumInstances,
      );
    }

    this.id = azurermKustoCluster.id;
    this.uri = azurermKustoCluster.uri;

    // Outputs
    const cdktfTerraformOutputKustoId = new cdktf.TerraformOutput(
      this,
      "Kusto_id",
      {
        value: azurermKustoCluster.id,
      },
    );
    const cdktfTerraformOutputKustoUri = new cdktf.TerraformOutput(
      this,
      "Kusto_uri",
      {
        value: azurermKustoCluster.uri,
      },
    );
    const cdktfTerraformOutputDataIngestionUri = new cdktf.TerraformOutput(
      this,
      "Kusto_data_ingestion_uri",
      {
        value: azurermKustoCluster.dataIngestionUri,
      },
    );
    const cdktfTerraformOutputKustoIdentity = new cdktf.TerraformOutput(
      this,
      "Kusto_identity",
      {
        value: azurermKustoCluster.identity,
        sensitive: true,
      },
    );

    /*This allows the Terraform resource name to match the original name. You can remove the call if you don't need them to match.*/
    cdktfTerraformOutputKustoId.overrideLogicalId("Kusto_id");
    cdktfTerraformOutputKustoUri.overrideLogicalId("Kusto_uri");
    cdktfTerraformOutputDataIngestionUri.overrideLogicalId(
      "Kusto_data_ingestion_uri",
    );
    cdktfTerraformOutputKustoIdentity.overrideLogicalId("Kusto_identity");
  }

  public addDatabase(databaseProps: DatabaseProps) {
    return new Database(this, databaseProps.name, databaseProps);
  }
}
