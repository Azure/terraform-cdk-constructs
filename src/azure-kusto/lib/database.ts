import { KustoDatabase } from "@cdktf/provider-azurerm/lib/kusto-database";
import { KustoDatabasePrincipalAssignment } from "@cdktf/provider-azurerm/lib/kusto-database-principal-assignment";
import { KustoScript } from "@cdktf/provider-azurerm/lib/kusto-script";
import * as cdktf from "cdktf";
import { Construct } from "constructs";
import { Md5 } from "ts-md5";
import { ClusterProps, Cluster } from "./cluster";

export interface DatabaseProps {
  /**
   * The Azure Kusto cluster to which this database belongs.
   */
  readonly kusto: Cluster;
  /**
   * The name of the Kusto Database to create.
   */
  readonly name: string;
  /**
   * The time the data that should be kept in cache for fast queries as ISO 8601 timespan.
   * Default is unlimited.
   */
  readonly hotCachePeriod?: string;
  /**
   * The time the data should be kept before it stops being accessible to queries as ISO 8601 timespan.
   * Default is unlimited.
   */
  readonly softDeletePeriod?: string;
}

export interface DatabaseAccessProps {
  /**
   * The name of the kusto principal assignment.
   */
  readonly name: string;
  /**
   * The tenant id in which the principal resides.
   */
  readonly tenantId: string;
  /**
   * The object id of the principal to assign to Kusto Database.
   */
  readonly principalId: string;
  /**
   * The type of the principal. Valid values include App, Group, User.
   */
  readonly principalType: string;
  /**
   * The database role assigned to the principal.
   * Valid values include Admin, Ingestor, Monitor, UnrestrictedViewer, User and Viewer.
   */
  readonly role: string;
}

export interface TableSchemaProps {
  readonly columnName: string;
  readonly columnType: string;
}

export class Database extends Construct {
  public readonly kustoDbProps: DatabaseProps;
  public readonly kustoProps: ClusterProps;
  public readonly rg: string;
  public readonly id: string;

  /**
   * Represents a Kusto Database within an Azure Kusto Cluster.
   *
   * This class is responsible for the creation and management of a database in Azure Data Explorer (Kusto),
   * which stores data tables and provides a query engine. A Kusto database is a logical group of tables
   * and is associated with a specific Kusto cluster. The database supports configurations such as
   * hot cache period and soft delete period to optimize performance and data retention according to
   * specific workload requirements.
   *
   * @param scope - The scope in which to define this construct, typically representing the Cloud Development Kit (CDK) stack.
   * @param id - The unique identifier for this instance of the Kusto database.
   * @param kustoDbProps - The properties required to configure the Kusto database. These include:
   *                       - `kusto`: Reference to the Kusto cluster to which the database will belong.
   *                       - `name`: The name of the database to be created within the Kusto cluster.
   *                       - `hotCachePeriod`: Optional. Specifies the duration that data should be kept in cache for faster query performance.
   *                                           Expressed in ISO 8601 duration format.
   *                       - `softDeletePeriod`: Optional. Specifies the duration that data should be retained before it stops being accessible.
   *                                             Expressed in ISO 8601 duration format.
   *
   * Example usage:
   * ```typescript
   * const myDatabase = new Database(this, 'MyKustoDatabase', {
   *   kusto: myKustoCluster,
   *   name: 'AnalyticsDB',
   *   hotCachePeriod: 'P30D',  // 30 days
   *   softDeletePeriod: 'P365D' // 365 days
   * });
   * ```
   *
   * This class sets up the database configurations and integrates it within the specified Kusto cluster,
   * providing capabilities to manage and query large datasets effectively.
   */
  constructor(scope: Construct, id: string, kustoDbProps: DatabaseProps) {
    super(scope, id);
    this.kustoDbProps = kustoDbProps;
    this.kustoProps = kustoDbProps.kusto.props;
    this.rg = kustoDbProps.kusto.resourceGroup.name;

    const kustoDatabase = new KustoDatabase(
      this,
      `kusto-db-${this.kustoDbProps.name}`,
      {
        name: this.kustoDbProps.name,
        location: kustoDbProps.kusto.resourceGroup.location,
        resourceGroupName: kustoDbProps.kusto.resourceGroup.name,
        clusterName: kustoDbProps.kusto.props.name,
      },
    );

    if (this.kustoDbProps.hotCachePeriod) {
      kustoDatabase.addOverride(
        "hot_cache_period",
        this.kustoDbProps.hotCachePeriod,
      );
    }
    if (this.kustoDbProps.softDeletePeriod) {
      kustoDatabase.addOverride(
        "soft_delete_period",
        this.kustoDbProps.softDeletePeriod,
      );
    }

    // Outputs
    this.id = kustoDatabase.id;
    const cdktfTerraformOutputKustoDbId = new cdktf.TerraformOutput(
      this,
      "id",
      {
        value: this.id,
      },
    );
    cdktfTerraformOutputKustoDbId.overrideLogicalId("id");
  }

  /**
   * Adds a database principal assignment in the Kusto cluster, assigning specified access rights to a principal.
   *
   * This method is used to grant access permissions to a specific user, group, or service principal within an Azure Active Directory.
   * These permissions determine the level of access that the principal has over the Kusto database, such as viewing, ingesting, or managing data.
   * The assignment is made by creating a KustoDatabasePrincipalAssignment resource, specifying the principal details and the type of role
   * they should assume.
   *
   * @param kustoDatabaseAccessProps - The properties defining the principal assignment. This includes:
   *   - `name`: A unique name identifying this principal assignment.
   *   - `tenantId`: The Azure Active Directory tenant ID where the principal resides.
   *   - `principalId`: The object ID of the principal (user, group, or service principal) in Azure AD.
   *   - `principalType`: The type of principal (e.g., User, Group, App).
   *   - `role`: The database role assigned to the principal. Roles can include Admin, User, Viewer, among others.
   *
   * Example usage:
   * ```typescript
   * myDatabase.addPermission({
   *   name: 'DataScienceTeamAccess',
   *   tenantId: 'tenant-id',
   *   principalId: 'principal-id',
   *   principalType: 'Group',
   *   role: 'Viewer'
   * });
   * ```
   * This method creates a new principal assignment, enabling specified access controls for the principal
   * on the Kusto database based on the role assigned. It is crucial for managing security and access
   * governance within the Kusto environment.
   */
  public addPermission(kustoDatabaseAccessProps: DatabaseAccessProps) {
    new KustoDatabasePrincipalAssignment(
      this,
      `kusto-db-${kustoDatabaseAccessProps.name}-access`,
      {
        name: kustoDatabaseAccessProps.name,
        resourceGroupName: this.rg,
        clusterName: this.kustoProps.name,
        databaseName: this.kustoDbProps.name,
        tenantId: kustoDatabaseAccessProps.tenantId,
        principalId: kustoDatabaseAccessProps.principalId,
        principalType: kustoDatabaseAccessProps.principalType,
        role: kustoDatabaseAccessProps.role,
      },
    );
  }

  /**
   * Adds a new table to an existing Azure Kusto database.
   *
   * This method creates a table within the specified Kusto database using a given schema. Tables in Kusto store structured data with
   * defined columns and types, which are crucial for storing and querying large datasets efficiently. The method constructs a Kusto
   * Data Explorer control command to create the table and then executes this command within the context of the database.
   *
   * @param tableName - The name of the table to create, which must be unique within the database.
   * @param tableSchema - An array of schema properties defining the columns of the table, including column names and their data types.
   *
   * Example usage:
   * ```typescript
   * myDatabase.addTable('SalesData', [
   *   { columnName: 'TransactionId', columnType: 'int' },
   *   { columnName: 'TransactionDate', columnType: 'datetime' },
   *   { columnName: 'Amount', columnType: 'real' }
   * ]);
   * ```
   * This method constructs the command to create the table and applies it directly within the Kusto database,
   * ensuring the table is ready for data ingestion and querying.
   */
  public addTable(tableName: string, tableSchema: TableSchemaProps[]) {
    const schemaContent = tableSchema
      .map((column) => {
        return `${column.columnName}:${column.columnType}`;
      })
      .join(", ");
    const scriptContent = `.create table ${tableName} ( ${schemaContent} )`;

    new KustoScript(
      this,
      `kusto-db-${this.kustoDbProps.name}-table-${tableName}`,
      {
        name: tableName,
        databaseId: this.id,
        scriptContent: scriptContent,
        continueOnErrorsEnabled: false,
        forceAnUpdateWhenValueChanged: Md5.hashStr(scriptContent),
      },
    );
  }

  /**
   * Adds and executes a control command or script within the Kusto database.
   *
   * This method facilitates the execution of Kusto Query Language (KQL) scripts or control commands within the specified
   * Kusto database. Scripts can perform a variety of functions, from schema modifications, like adding new tables or altering
   * existing ones, to data management operations, such as data ingestion or cleanup tasks. Each script is executed as a
   * KustoScript resource, which ensures that the script is applied correctly and atomically to the database.
   *
   * @param scriptName - A unique name for the script, which helps in identifying the script resource within the deployment.
   * @param scriptContent - The KQL script or control command to be executed. This should be a valid KQL command string.
   *
   * Example usage:
   * ```typescript
   * myDatabase.addScript('InitializeSalesTable', `
   *   .create table SalesData (TransactionId: int, TransactionDate: datetime, Amount: real)
   *   .alter-merge table SalesData policy retentionsoftdelete = 365d
   * `);
   * ```
   * This method will create a `KustoScript` resource that encapsulates the command, ensuring it is executed against the
   * database, and is tracked as part of the resource management within Azure.
   */
  public addScript(scriptName: string, scriptContent: string) {
    new KustoScript(
      this,
      `kusto-db-${this.kustoDbProps.name}-script-${scriptName}`,
      {
        name: `script-${scriptName}`,
        databaseId: this.id,
        scriptContent: scriptContent,
        continueOnErrorsEnabled: false,
        forceAnUpdateWhenValueChanged: Md5.hashStr(scriptContent),
      },
    );
  }
}
