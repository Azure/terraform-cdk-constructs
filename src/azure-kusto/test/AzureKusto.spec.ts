import { AzurermProvider } from "@cdktf/provider-azurerm/lib/provider";
import { Testing, TerraformStack } from "cdktf";
import "cdktf/lib/testing/adapters/jest";
import { exampleAzureKusto } from "./ExampleAzureKusto";
import * as kusto from "..";

describe("Kusto With Defaults", () => {
  let stack: TerraformStack;
  let fullSynthResult: any;

  beforeEach(() => {
    jest.mock("../../azure-resourcegroup", () => ({
      Group: {
        Location: "eastus",
        Name: "testrgname",
      },
    }));
    const rgMock = jest.requireMock("../../azure-resourcegroup");

    const app = Testing.app();
    stack = new TerraformStack(app, "test");

    new AzurermProvider(stack, "azureFeature", { features: {} });
    new kusto.Cluster(stack, "testAzureKustoDefaults", {
      rg: rgMock.Group,
      name: "kustotest",
      sku: kusto.ComputeSpecification.devtestExtraSmallEav4,
    });

    fullSynthResult = Testing.fullSynth(stack); // Save the result for reuse
  });

  it("renders an Kusto with defaults and checks snapshot", () => {
    expect(Testing.synth(stack)).toMatchSnapshot(); // Compare the already prepared stack
  });

  it("check if the produced terraform configuration is valid", () => {
    expect(fullSynthResult).toBeValidTerraform(); // Use the saved result
  });

  it("check if this can be planned", () => {
    expect(fullSynthResult).toPlanSuccessfully(); // Use the saved result
  });
});

describe("Kusto Example", () => {
  it("renders the Azure Kusto Example and checks snapshot", () => {
    expect(
      Testing.synth(new exampleAzureKusto(Testing.app(), "testAzureKusto")),
    ).toMatchSnapshot();
  });

  it("check if the produced terraform configuration is valid", () => {
    // We need to do a full synth to plan the terraform configuration
    expect(
      Testing.fullSynth(new exampleAzureKusto(Testing.app(), "testAzureKusto")),
    ).toBeValidTerraform();
  });

  it("check if this can be planned", () => {
    // We need to do a full synth to plan the terraform configuration
    expect(
      Testing.fullSynth(new exampleAzureKusto(Testing.app(), "testAzureKusto")),
    ).toPlanSuccessfully();
  });
});
