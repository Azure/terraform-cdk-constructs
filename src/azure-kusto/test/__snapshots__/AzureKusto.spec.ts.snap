// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kusto Example renders the Azure Kusto Example and checks snapshot 1`] = `
"{
  "data": {
    "azurerm_client_config": {
      "CurrentClientConfig": {
      }
    }
  },
  "output": {
    "Kusto_data_ingestion_uri": {
      "value": "\${azurerm_kusto_cluster.kusto_Kusto_6B093AC5.data_ingestion_uri}"
    },
    "Kusto_id": {
      "value": "\${azurerm_kusto_cluster.kusto_Kusto_6B093AC5.id}"
    },
    "Kusto_identity": {
      "sensitive": true,
      "value": "\${azurerm_kusto_cluster.kusto_Kusto_6B093AC5.identity[0]}"
    },
    "Kusto_uri": {
      "value": "\${azurerm_kusto_cluster.kusto_Kusto_6B093AC5.uri}"
    },
    "id": {
      "value": "\${azurerm_kusto_database.kusto_testDB1_kusto-db-testDB1_D1A53426.id}"
    },
    "location": {
      "value": "\${azurerm_resource_group.rg_1429A41A.location}"
    },
    "name": {
      "value": "\${azurerm_resource_group.rg_1429A41A.name}"
    }
  },
  "provider": {
    "azurerm": [
      {
        "features": {
          "resource_group": {
            "prevent_deletion_if_contains_resources": false
          }
        }
      }
    ]
  },
  "resource": {
    "azurerm_kusto_cluster": {
      "kusto_Kusto_6B093AC5": {
        "auto_stop_enabled": true,
        "identity": {
          "identity_ids": [
          ],
          "type": "SystemAssigned"
        },
        "location": "\${azurerm_resource_group.rg_1429A41A.location}",
        "name": "kusto\${var.name}",
        "public_network_access_enabled": true,
        "purge_enabled": false,
        "resource_group_name": "\${azurerm_resource_group.rg_1429A41A.name}",
        "sku": {
          "capacity": 2,
          "name": "Dev(No SLA)_Standard_D11_v2"
        },
        "streaming_ingestion_enabled": true,
        "zones": [
          "1",
          "2",
          "3"
        ]
      }
    },
    "azurerm_kusto_database": {
      "kusto_testDB1_kusto-db-testDB1_D1A53426": {
        "cluster_name": "kusto\${var.name}",
        "hot_cache_period": "P7D",
        "location": "\${azurerm_resource_group.rg_1429A41A.location}",
        "name": "testDB1",
        "resource_group_name": "\${azurerm_resource_group.rg_1429A41A.name}",
        "soft_delete_period": "P31D"
      }
    },
    "azurerm_kusto_database_principal_assignment": {
      "kusto_testDB1_kusto-db-User1Admin-access_836A7F36": {
        "cluster_name": "kusto\${var.name}",
        "database_name": "testDB1",
        "name": "User1Admin",
        "principal_id": "\${data.azurerm_client_config.CurrentClientConfig.client_id}",
        "principal_type": "User",
        "resource_group_name": "\${azurerm_resource_group.rg_1429A41A.name}",
        "role": "Admin",
        "tenant_id": "\${data.azurerm_client_config.CurrentClientConfig.tenant_id}"
      }
    },
    "azurerm_kusto_script": {
      "kusto_testDB1_kusto-db-testDB1-script-MyTestScript_228E71EF": {
        "continue_on_errors_enabled": false,
        "database_id": "\${azurerm_kusto_database.kusto_testDB1_kusto-db-testDB1_D1A53426.id}",
        "force_an_update_when_value_changed": "1b98d80c44453d043e9faf8aa4ca9e9e",
        "name": "script-MyTestScript",
        "script_content": ".create table MyTestTable2 ( Timestamp:datetime, User:string, Value:int32 )"
      },
      "kusto_testDB1_kusto-db-testDB1-table-MyTestTable_A0AAD29F": {
        "continue_on_errors_enabled": false,
        "database_id": "\${azurerm_kusto_database.kusto_testDB1_kusto-db-testDB1_D1A53426.id}",
        "force_an_update_when_value_changed": "9b2184ae8b9df7cda6deb740fbba4e37",
        "name": "MyTestTable",
        "script_content": ".create table MyTestTable ( Timestamp:datetime, User:string, Value:int32 )"
      }
    },
    "azurerm_resource_group": {
      "rg_1429A41A": {
        "lifecycle": [
          {
            "ignore_changes": [
            ]
          }
        ],
        "location": "eastus",
        "name": "rg-\${var.name}"
      }
    },
    "azurerm_role_assignment": {
      "kusto_TfTokenTOKEN47Contributor_role_1777E8D7": {
        "principal_id": "\${data.azurerm_client_config.CurrentClientConfig.object_id}",
        "role_definition_name": "Contributor",
        "scope": "\${azurerm_kusto_cluster.kusto_Kusto_6B093AC5.id}"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.75.0"
      }
    }
  },
  "variable": {
    "name": {
      "default": "test",
      "description": "System name used to randomize the resources",
      "type": "string"
    }
  }
}"
`;

exports[`Kusto With Defaults renders an Kusto with defaults and checks snapshot 1`] = `
"{
  "output": {
    "Kusto_data_ingestion_uri": {
      "value": "\${azurerm_kusto_cluster.testAzureKustoDefaults_Kusto_4306B42C.data_ingestion_uri}"
    },
    "Kusto_id": {
      "value": "\${azurerm_kusto_cluster.testAzureKustoDefaults_Kusto_4306B42C.id}"
    },
    "Kusto_identity": {
      "sensitive": true,
      "value": "\${azurerm_kusto_cluster.testAzureKustoDefaults_Kusto_4306B42C.identity[0]}"
    },
    "Kusto_uri": {
      "value": "\${azurerm_kusto_cluster.testAzureKustoDefaults_Kusto_4306B42C.uri}"
    }
  },
  "provider": {
    "azurerm": [
      {
        "features": {
        }
      }
    ]
  },
  "resource": {
    "azurerm_kusto_cluster": {
      "testAzureKustoDefaults_Kusto_4306B42C": {
        "auto_stop_enabled": true,
        "identity": {
          "identity_ids": [
          ],
          "type": "SystemAssigned"
        },
        "location": "eastus",
        "name": "kustotest",
        "public_network_access_enabled": true,
        "purge_enabled": false,
        "resource_group_name": "testrgname",
        "sku": {
          "capacity": 2,
          "name": "Dev(No SLA)_Standard_D11_v2"
        },
        "streaming_ingestion_enabled": true,
        "zones": [
          "1",
          "2",
          "3"
        ]
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.75.0"
      }
    }
  }
}"
`;
